opensearch/src/main/java/org/opensearch/sql/opensearch/planner/physical/ADOperator.java

```diff
@@ -0,0 +1,213 @@
+            MLClient.getMLClient(nodeClient);
+    MLPredictionOutput predictionResult = (MLPredictionOutput) machineLearningClient
+            .trainAndPredict(mlinput)
+            .actionGet(30, TimeUnit.SECONDS);
+    Iterator<Row> inputRowIter = inputDataFrame.iterator();
+    Iterator<Row> resultRowIter = predictionResult.getPredictionResult().iterator();
+    iterator = new Iterator<ExprValue>() {
+      @Override
+      public boolean hasNext() {
+        return inputRowIter.hasNext();
+      }
+
+      @Override
+      public ExprValue next() {
+        ImmutableMap.Builder<String, ExprValue> resultSchemaBuilder = new ImmutableMap.Builder<>();
+        resultSchemaBuilder.putAll(convertRowIntoExprValue(inputDataFrame.columnMetas(),
+                inputRowIter.next()));
+        Map<String, ExprValue> resultSchema = resultSchemaBuilder.build();
+        ImmutableMap.Builder<String, ExprValue> resultBuilder = new ImmutableMap.Builder<>();
+        resultBuilder.putAll(convertResultRowIntoExprValue(
+                predictionResult.getPredictionResult().columnMetas(),
+                resultRowIter.next(),
+                resultSchema));
+        resultBuilder.putAll(resultSchema);
+        return ExprTupleValue.fromExprValueMap(resultBuilder.build());
+      }
+    };
+  }
+
+  @Override
+  public <R, C> R accept(PhysicalPlanNodeVisitor<R, C> visitor, C context) {
+    return visitor.visitAD(this, context);
+  }
+
+  @Override
+  public boolean hasNext() {
+    return iterator.hasNext();
+  }
+
+  @Override
+  public ExprValue next() {
+    return iterator.next();
+  }
+
+  @Override
+  public List<PhysicalPlan> getChild() {
+    return Collections.singletonList(input);
+  }
+
+  protected MLAlgoParams convertArgumentToMLParameter(Map<String, Literal> arguments) {
+    if (arguments.get("time_field").getValue() == null) {
+      rcfType = FunctionName.BATCH_RCF;
+      return BatchRCFParams.builder()
+              .shingleSize((Integer) arguments.get("shingle_size").getValue())
+              .build();
+    }
+    rcfType = FunctionName.FIT_RCF;
+    return FitRCFParams.builder()
+            .shingleSize((Integer) arguments.get("shingle_size").getValue())
+            .timeDecay((Double) arguments.get("time_decay").getValue())
+            .timeField((String) arguments.get("time_field").getValue())
+            .dateFormat("yyyy-MM-dd HH:mm:ss")
+            .build();
+  }
+
+  private Map<String, ExprValue> convertRowIntoExprValue(ColumnMeta[] columnMetas, Row row) {
+    ImmutableMap.Builder<String, ExprValue> resultBuilder = new ImmutableMap.Builder<>();
+    for (int i = 0; i < columnMetas.length; i++) {
+      ColumnValue columnValue = row.getValue(i);
+      String resultKeyName = columnMetas[i].getName();
+      popluateResultBuilder(columnValue, resultKeyName, resultBuilder);
+    }
+    return resultBuilder.build();
+  }
+
+  private void popluateResultBuilder(ColumnValue columnValue,
```

Comment:
Is it duplicated with convertRowIntoExprValue() in MLCommonsOperator?

Reply:
I think we can consider abstract such common shared code, that may benefit general `train` , `predict` and `train_predict` command

Reply:
Abstract duplicate code in the latest revision